src/coprocessor/codec/convert.rs

```diff
@@ -678,117 +680,168 @@ fn round_int_str(num_next_dot: char, s: &str) -> Cow<'_, str> {
 /// because precision will be lost.
 ///
 /// When the float string indicating a value that is overflowing the i64,
-/// the original float string is returned and an overflow warning is attached
+/// the original float string is returned and an overflow warning is attached.
+///
+/// This func will find serious overflow such as the len of result > 20 (without prefix `+/-`)
+/// however, it will not check whether the result overflow BIGINT.
 fn float_str_to_int_string<'a>(
     ctx: &mut EvalContext,
     valid_float: &'a str,
 ) -> Result<Cow<'a, str>> {
+    // this func is complex, to make it same as TiDB's version,
+    // we impl it like TiDB's version(https://github.com/pingcap/tidb/blob/9b521342bf/types/convert.go#L400)
     let mut dot_idx = None;
     let mut e_idx = None;
-    let mut int_cnt: i64 = 0;
-    let mut digits_cnt: i64 = 0;
 
     for (i, c) in valid_float.chars().enumerate() {
         match c {
             '.' => dot_idx = Some(i),
             'e' | 'E' => e_idx = Some(i),
-            '0'..='9' => {
-                if e_idx.is_none() {
-                    if dot_idx.is_none() {
-                        int_cnt += 1;
-                    }
-                    digits_cnt += 1;
-                }
-            }
             _ => (),
         }
     }
 
     if dot_idx.is_none() && e_idx.is_none() {
```

Comment:
The more idiom way:
```rust
    if dot_idx.is_none() && e_idx.is_none() {
        return Ok(Cow::Borrowed(valid_float));
    }
    
    if e_idx.is_none() {
        no_exp_float_str_to_int_str(valid_float, dot_idx.unwrap())
    } else {
        exp_float_str_to_int_str(ctx, valid_float, e_idx.unwrap(), dot_idx)
    }
```

Reply:
The most idiomatic way is to use pattern matching.

```rust
match (dot_idx, e_idx) {
    (None, None) => return Ok(Cow::Borrowed(valid_float)),
    (Some(di), None) => no_exp_float_str_to_int_str(valid_float, di),
    (di, Some(ei)) => exp_float_str_to_int_str(ctx, valid_float, ei, di),
}
```